src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -219,73 +249,84 @@ static void CheckForExperimentStatus(ITextView textView, Document document)
+
             if (completionList == null)
             {
-                return new AsyncCompletionData.CompletionContext(ImmutableArray<VSCompletionItem>.Empty);
+                items = ImmutableArray<VSCompletionItem>.Empty;
+                suggestionItemOptions = null;
             }
-
-            var itemsBuilder = new ArrayBuilder<VSCompletionItem>(completionList.Items.Length);
-            foreach (var roslynItem in completionList.Items)
+            else
             {
-                cancellationToken.ThrowIfCancellationRequested();
-                var item = Convert(document, roslynItem);
-                itemsBuilder.Add(item);
-            }
-
-            var items = itemsBuilder.ToImmutableAndFree();
+                var itemsBuilder = new ArrayBuilder<VSCompletionItem>(completionList.Items.Length);
+                foreach (var roslynItem in completionList.Items)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var item = Convert(document, roslynItem, filterSet);
+                    itemsBuilder.Add(item);
+                }
 
-            var suggestionItemOptions = completionList.SuggestionModeItem != null
-                    ? new AsyncCompletionData.SuggestionItemOptions(
-                        completionList.SuggestionModeItem.DisplayText,
-                        completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
-                            ? description
-                            : string.Empty)
-                    : null;
+                items = itemsBuilder.ToImmutableAndFree();
 
-            // Have to store the snapshot to reuse it in some projections related scenarios
-            // where data and session in further calls are able to provide other snapshots.
-            session.Properties.AddProperty(TriggerSnapshot, triggerLocation.Snapshot);
+                suggestionItemOptions = completionList.SuggestionModeItem != null
+                        ? new AsyncCompletionData.SuggestionItemOptions(
+                            completionList.SuggestionModeItem.DisplayText,
+                            completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
+                                ? description
+                                : string.Empty)
+                        : null;
 
-            // Store around the span this completion list applies to.  We'll use this later
-            // to pass this value in when we're committing a completion list item.
-            session.Properties.AddProperty(CompletionListSpan, completionList.Span);
+                if (!isExpanded)
+                {
+                    // Store around the span this completion list applies to.  We'll use this later
+                    // to pass this value in when we're committing a completion list item.
+                    session.Properties[CompletionListSpan] = completionList.Span;
```

Comment:
>[CompletionListSpan] [](start = 38, length = 20)

So, now this definitely can be called more than once. Then, we should replace AddProperty with Set. 

There are some more points to consider (this applies to all property bag sets):
1. Different providers may call this within the same session, such as TS vs C#. With the current approach, the latest one will win.
2. Some settings maybe have to be set on the first pass, and maybe should not be updated.

We have to review each of those assignments and decide what would be the right approach for it: 
1. override, the latest wins
2. merge
3. keep separate settings per content type
etc.

If you think the current PR is not a good place for this, please create an issue, and add TODOs for each property bag assignment.

Reply:
I have made the change to merge those values as discussed offline. Will do `keep separate settings per content type` in a separate PR.

Reply:
Sounds good! Could you please add a TODO here with a GitHub issue number to track?

---
In reply to: [296063063](https://github.com/dotnet/roslyn/pull/36146#discussion_r296063063) [](ancestors = 296063063)